### PR TITLE
Refactor LookupTable and LookupRowFinder into a single class

### LINKED ISSUE #33512: Match lookup rows using lookup text when a user processes a row
**Refactoring**

- [x] Refactor the contents of `Experiment::findLookupRow` into a separate class e.g. `LookupRowFinder` and add tests for it
  - [x] Test lookup of exact theta match works
  - [x] Test lookup of different theta but within tolerance works
  - [x] Test lookup of theta outside tolerance fails and uses wildcard row instead
  - [x] Test lookup of theta fails and there is no wildcard row does not match anything
  - [x] Test lookup with empty table
- [x] Refactor the method to take a `Row` instead of the angle
- [x] #33546
- [x] Merge the `LookupRowFinder` method into `LookupTable`

**New feature**

- [x] Add a new field for the regex expression to the `LookupRow` model
- [x] Adjust the `LookupTable` to use the new regex field as well as the angle when looking up a row
  - [x] If both angle and regex are specified then check both. Ensure we do this before checking wildcard rows or rows where the regex is not set.
  - [x] Secondly, if the angle is specified but the regex is not then just check the angle
  - [x] Thirdly, check wildcard rows (where neither the angle or regex is specified)

### LINKED ISSUE #33546: Create a real class for LookupTable
In the ISIS Reflectometry interface, currently `LookupTable` is just a vector of `LookupRow`s. We now need to add more functionality to it so it should be converted into a real class. The functionality from `LookupRowFinder` should then be merged into it and the `LookupRowFinder` class removed.

### PR DESCRIPTION
**Pull request by gemmaguest**
This PR changes `LookupTable` from a vector of rows to its own class, and merges the functionality from `LookupRowFinder` into it. 

**To test:**

Follow the instructions on #33535 to ensure existing functionality is not broken

Fixes #33546 and fixes #33512

*This does not require release notes* because **it is under-the-hood work that does not change functionality**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gemmaguest** [comment]
@david This is ready for review now

File: qt/scientific_interfaces/ISISReflectometry/Reduction/LookupTable.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+std::vector<LookupRow::ValueArray> LookupTable::toValueArray() const {
+  auto result = std::vector<LookupRow::ValueArray>();
+  std::transform(m_lookupRows.cbegin(), m_lookupRows.cend(), std::back_inserter(result),
+                 [](auto const &lookupRow) { return lookupRowToArray(lookupRow); });
+  return result;
+};
+
+bool operator==(LookupTable const &lhs, LookupTable const &rhs) { return lhs.m_lookupRows == rhs.m_lookupRows; }
+bool operator!=(LookupTable const &lhs, LookupTable const &rhs) { return !(lhs == rhs); }
```
Comment (DavidFair) [inline qt/scientific_interfaces/ISISReflectometry/Reduction/LookupTable.cpp:108]: So I looked this up as the current way looked wrong to me (despite writing it originally), apparently the idiomatic way of doing this is 
```c++
return !operator==(lhs, rhs);
```

File: qt/scientific_interfaces/ISISReflectometry/test/Experiment/ExperimentOptionDefaultsTest.h [hunk 2]
```diff
@@ -74,16 +74,20 @@ class ExperimentOptionDefaultsTest : public CxxTest::TestSuite {
     auto result = getDefaults();
     auto expected = LookupRow(boost::none, boost::none, TransmissionRunPair(), boost::none,
                               RangeInQ(boost::none, boost::none, boost::none), boost::none, boost::none, boost::none);
-    TS_ASSERT_EQUALS(result.lookupTable().size(), 1);
-    TS_ASSERT_EQUALS(result.lookupTable().front(), expected);
+    auto foundLookupRows = result.lookupTableRows();
+    TS_ASSERT_EQUALS(foundLookupRows.size(), 1);
+    if (foundLookupRows.size() > 0)
```
Comment (DavidFair) [inline qt/scientific_interfaces/ISISReflectometry/test/Experiment/ExperimentOptionDefaultsTest.h:79]: `!foundLookupRows.empty()`

**Review by DavidFair — CHANGES_REQUESTED** [review]
Some minor nits

**Comment by DavidFair** [comment]
Functionality wise, this worked correctly. I tested:
- Two angles and dq/q specified
- A wildcard row
- No match row

**Review by DavidFair — DISMISSED** [review]
LGTM let's :shipit:

**Review by DavidFair — APPROVED** [review]

**Pull request merged**